src/core/compiler/CircomCompilerFactory.ts

```diff
@@ -1,44 +1,129 @@
+  ): Promise<ICircomCompiler | IWASMCircomCompiler> {
+    if (!isVersionHigherOrEqual("2.1.9", highestVersion)) {
+      throw new HardhatZKitError(
+        `Unsupported Circom compiler version - ${highestVersion}. Please provide another version.`,
+      );
+    }
+
+    const nativeCompiler = await this._getNativeCompiler();
+
+    if (nativeCompiler && isVersionHigherOrEqual(nativeCompiler.version, highestVersion)) {
+      return new BinaryCircomCompiler(nativeCompiler.binaryPath);
     }
+
+    const { compilerPath, isWasm } = await this._getCircomCompiler(highestVersion);
+
+    if (isWasm) {
+      return new WASMCircomCompiler(this._getCircomWasmCompiler(compilerPath));
+    }
+
+    return new BinaryCircomCompiler(compilerPath);
   }
 
-  public static async checkNativeCompilerExistence() {
+  private static async _getNativeCompiler(): Promise<NativeCompiler | undefined> {
+    const execP = promisify(exec);
+
     try {
-      await execCall("circom", ["--version"]);
-    } catch (e) {
-      throw new HardhatZKitError(
-        "Native Circom compiler was not found. Set the compiler globally or change the 'nativeCompiler' flag to false",
-      );
+      const { stdout: circomLocation } = await execP("whereis circom");
+
+      const trimmedBinaryPath = circomLocation.trim().split(" ");
+
+      if (trimmedBinaryPath.length !== 2) {
+        return undefined;
+      }
+
+      const nativeCircomBinaryPath = trimmedBinaryPath[1];
+
+      const { stdout: versionOutput } = await execP(`circom --version`);
+
+      const versionParts = versionOutput.trim().split(" ");
+      const version = versionParts[versionParts.length - 1];
+
+      return {
+        binaryPath: nativeCircomBinaryPath,
+        version,
+      };
+    } catch (error) {
+      return undefined;
+    }
+  }
+
+  private static async _getCircomCompiler(
+    highestVersion: string,
+    isWasm: boolean = false,
+  ): Promise<{ compilerPath: string; isWasm: boolean }> {
+    const compilersDir = await this._getCompilersDir();
+
+    const compilerPlatformBinary = isWasm
+      ? CompilerPlatformBinary.WASM
+      : CompilerDownloader.getCompilerPlatformBinary();
+    const downloader = CompilerDownloader.getConcurrencySafeDownloader(compilerPlatformBinary, compilersDir);
+
+    if (await downloader.isCompatibleCompilerDownloaded(highestVersion)) {
+      const compilerPath = await downloader.getLatestCompilerBinary();
+      return { compilerPath, isWasm: this._isCompilerWasm(compilerPlatformBinary) };
+    } else {
+      try {
+        await downloader.downloadLatestCompiler();
+
+        const compilerPath = await downloader.getLatestCompilerBinary();
+
+        return { compilerPath, isWasm: this._isCompilerWasm(compilerPlatformBinary) };
+      } catch (error: any) {
+        if (this._isCompilerWasm(compilerPlatformBinary)) {
+          throw new HardhatZKitError(error);
+        }
+
+        return this._getCircomCompiler(highestVersion, true);
+      }
```

Comment:
Ok, but why does this function handle its errors? Better to stick to OOP principles and let this function be responsible for one thing only. 

Move try/catch upper the call stack and call this function twice.